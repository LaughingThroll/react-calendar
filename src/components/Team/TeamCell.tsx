import React from 'react'
import classnames from 'classnames'

import { isWeekend } from '../../utils/date/index'

interface ITeamCell {
  date: Date
}

const TeamCell: React.FC<ITeamCell> = ({ date }) => {
  return (
    <td
      className={classnames({
        'calendar-body__cell': true,
        'cell-gray': isWeekend(date),
      })}
    ></td>
  )
}

export default React.memo(TeamCell)
