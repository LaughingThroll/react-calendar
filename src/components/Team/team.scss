.team {
  padding: 7px 5px 7px 19px;
}
