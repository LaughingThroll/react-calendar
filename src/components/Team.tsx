import React from "react"

import iconUsers from "./../assets/images/icons/users.svg"

import TeamCell from "./TeamCell"
import TeamMember from "./TeamMember"

import { THEMES } from "../constant"
import { formatDayInBinaryString } from "../utils/date"

import { IMember, ITeam } from "../types/DB"

interface ITeamComponent {
  team: ITeam
  date: Date
  allDaysInMonth: number
  themeIndex?: number
}

const Team: React.FC<ITeamComponent> = ({
  team: { members, percentageOfAbsent, name: nameTeam },
  allDaysInMonth,
  date,
  themeIndex = 0,
}) => {
  const mainTheme = THEMES[themeIndex % THEMES.length][0]
  return (
    <>
      <tr className={`calendar-team calendar-table--indentation ${mainTheme} ${THEMES[themeIndex % THEMES.length][1]}`}>
        <td className="team team--common">
          <span className="team__name">{nameTeam}</span>
          <div className="team__other">
            <div className="team__users users">
              <img className="users__svg" src={iconUsers} alt="" />
              <span className="users__count">{members.length}</span>
            </div>
            <div className="team__weekend-percent weekend-percent">${percentageOfAbsent[date.getMonth()]}</div>
            <button className="button-arrow-up"></button>
          </div>
        </td>
        {new Array(allDaysInMonth).fill(0).map((_, day) => (
          <TeamCell dayString={formatDayInBinaryString(date, day + 1)} />
        ))}
        <td className="calendar-team__cell cell-gray"></td>
      </tr>
      {members.map(({ name: nameMember }: IMember) => (
        <TeamMember date={date} allDaysInMonth={allDaysInMonth} name={nameMember} theme={mainTheme} />
      ))}
    </>
  )
}

export default Team
