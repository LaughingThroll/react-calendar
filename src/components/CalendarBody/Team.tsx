import React, { useState } from 'react'

import iconUsers from './../../assets/images/icons/users.svg'

import TeamCell from './TeamCell'
import TeamMember from './TeamMember'

import { THEMES } from '../../constant'
import { formatDayInBinaryString } from '../../utils/date'

import { ITeam } from '../../types/model/team'
import { IMember } from '../../types/model/member'
// import { IMember, ITeam } from '../../types/DB'

interface ITeamComponent {
  team: ITeam
  date: Date
  allDaysInMonth: number
  themeIndex?: number
}

const Team: React.FC<ITeamComponent> = ({
  team: { members, percentageOfAbsent, name: nameTeam },
  allDaysInMonth,
  date,
  themeIndex = 0,
}) => {
  const mainTheme = THEMES[themeIndex % THEMES.length][0]
  const [isGroupOpen, setIsGrpupOpen] = useState(true)

  const handlerHideGroup = () => {
    setIsGrpupOpen(!isGroupOpen)
  }

  return (
    <>
      <tr
        className={`calendar-team calendar-table--indentation ${mainTheme} ${THEMES[themeIndex % THEMES.length][1]} ${
          isGroupOpen ? '' : 'is-group-closed'
        }`}
      >
        <td className="team team--common">
          <span className="team__name">{nameTeam}</span>
          <div className="team__other">
            <div className="team__users users">
              <img className="users__svg" src={iconUsers} alt="" />
              <span className="users__count">{members.length}</span>
            </div>
            <div className="team__weekend-percent weekend-percent">{percentageOfAbsent[date.getMonth()]}%</div>
            <button
              className={`button-arrow-up ${isGroupOpen ? 'is-group-open-btn' : 'is-group-closed-btn'}`}
              onClick={() => handlerHideGroup()}
            ></button>
          </div>
        </td>
        {new Array(allDaysInMonth).fill(0).map((_, day: number) => (
          <TeamCell key={day} dayString={formatDayInBinaryString(date, day + 1)} />
        ))}
        <td className="calendar-team__cell cell-gray"></td>
      </tr>
      {members.map(({ name: nameMember, vacations: arrayOfVacations }: IMember, day: number) => (
        <TeamMember
          key={day}
          date={date}
          allDaysInMonth={allDaysInMonth}
          name={nameMember}
          vacations={arrayOfVacations}
          theme={mainTheme}
          isGroupOpen={isGroupOpen}
        />
      ))}
    </>
  )
}

export default Team
