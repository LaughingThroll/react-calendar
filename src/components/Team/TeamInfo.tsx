import React from 'react'
import classnames from 'classnames'

import { useToggle } from '../../hooks'

import iconUsers from './../../assets/images/icons/users.svg'

import { ITeam } from '../../types/model/team'

interface ITeamInfo extends ITeam {
  date: Date
}

const TeamInfo: React.FC<ITeamInfo> = ({ date, name, members, percentageOfAbsent }) => {
  const { isOpen, changeOpen } = useToggle()

  const getPercentageOfAbsentCount = (percentageOfAbsent: number[], date: Date): number => {
    return date.getFullYear() === new Date().getFullYear() ? percentageOfAbsent[date.getMonth()] : 0
  }

  return (
    <div className="team-info">
      <span className="team-info__name">{name}</span>
      <div className="team-info__other">
        <div className="team-info__users users">
          <img className="users__svg" src={iconUsers} alt="" />
          <span className="users__count">{members.length}</span>
        </div>
        <div className="team-info__weekend-percent weekend-percent">
          {getPercentageOfAbsentCount(percentageOfAbsent, date)}%
        </div>
        <button
          className={classnames({
            'button-arrow-up': true,
            'is-group-open-btn': isOpen,
            'is-group-closed-btn': !isOpen,
          })}
          onClick={changeOpen}
        ></button>
      </div>
    </div>
  )
}

export default TeamInfo
