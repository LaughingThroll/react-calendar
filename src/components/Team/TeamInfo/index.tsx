import React from 'react'
import classnames from 'classnames'

import styles from './teamInfo.module.scss'

import { CountUsers, WeekendPercents } from '../../common'

import { getPercentageOfAbsentCount } from '../../../utils/teams'

import { Team } from '../../../types/model/team'

export interface TeamInfoProps extends Team {
  date: Date
  isOpen?: boolean
  toggleOpen?: () => void
  classNames?: string[]
}

const { teamInfo, teamInfo__name, teamInfo__other, teamInfo__countUsers } = styles

const TeamInfo: React.FC<TeamInfoProps> = ({
  date,
  name,
  members,
  percentageOfAbsent,
  isOpen,
  toggleOpen,
  classNames = [],
}) => {
  return (
    <div className={[teamInfo, ...classNames].join(' ')}>
      <span className={teamInfo__name}>{name}</span>
      <div className={teamInfo__other}>
        <CountUsers countUsers={members.length} classNames={[teamInfo__countUsers]} />
        <WeekendPercents
          percent={getPercentageOfAbsentCount(percentageOfAbsent, date)}
          classNames={['teamInfo__weekendPercent']}
        />
        <button
          className={classnames({
            'button-arrow-up': true,
            'button-arrow-up--active': isOpen,
          })}
          onClick={toggleOpen}
        ></button>
      </div>
    </div>
  )
}

export default React.memo(TeamInfo)
