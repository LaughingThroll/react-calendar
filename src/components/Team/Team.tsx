import React from 'react'
import classnames from 'classnames'

import { Cell } from '../common'
import TeamInfo from './TeamInfo'
import Member from '../Member/Member'

import { useToggle } from '../../hooks'

import { THEMES } from '../../theme'
import { getAllDaysInMonth } from '../../utils/date'

import { Team } from '../../types/model/team'

export interface TeamProps {
  date: Date
  team: Team
  themeIndex?: number
  classNames?: string[]
}

const TeamComponent: React.FC<TeamProps> = ({ team, date, themeIndex = 0, classNames = [] }) => {
  const mainTheme = THEMES[themeIndex % THEMES.length][0]
  const { isOpen, toggleOpen } = useToggle(true)

  return (
    <>
      <tr
        className={`
        calendar-body__row 
        calendar-table--indentation 
        ${mainTheme} 
        ${THEMES[themeIndex % THEMES.length][1]} 
        ${classnames({ 'calendar-table--closed': !isOpen })} `}
      >
        <td className="team">
          <TeamInfo date={date} {...team} isOpen={isOpen} toggleOpen={toggleOpen} />
        </td>
        {getAllDaysInMonth(date).map((date, index) => (
          <Cell key={index} cellDate={date} classNames={['calendar-body__cell']}></Cell>
        ))}

        <td className="calendar-body__cell cell-gray"></td>
      </tr>

      {isOpen && team.members.map((member) => <Member key={member.id} date={date} theme={mainTheme} {...member} />)}
    </>
  )
}

export default React.memo(TeamComponent)
