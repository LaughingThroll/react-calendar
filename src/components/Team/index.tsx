import React from 'react'
import classnames from 'classnames'

import './team.scss'

import { Cell } from '../common'
import TeamInfo from './TeamInfo'
import Member from '../Member'

import { useToggle } from '../../hooks'

import { THEMES } from '../../theme'
import { getAllDaysInMonth } from '../../utils/date'

import { Team } from '../../types/model/team'

export interface TeamProps {
  date: Date
  team: Team
  themeIndex?: number
  classNames?: string[]
}

const TeamComponent: React.FC<TeamProps> = ({ team, date, themeIndex = 0, classNames = [] }) => {
  const mainTheme = THEMES[themeIndex % THEMES.length][0]
  const { isOpen, toggleOpen } = useToggle(true)

  return (
    <>
      <tr
        className={`
        calendar-body__row  
        ${mainTheme} 
        ${THEMES[themeIndex % THEMES.length][1]} 
        ${classnames({ 'calendar-table--closed': !isOpen })} ${[...classNames].join(' ')}`}
      >
        <td className="team">
          <TeamInfo date={date} {...team} isOpen={isOpen} toggleOpen={toggleOpen} />
        </td>
        {getAllDaysInMonth(date).map((date, index) => (
          <Cell key={index} cellDate={date} classNames={['calendar-body__cell']}></Cell>
        ))}

        <Cell classNames={['calendar-body__cell', 'cell-gray']}></Cell>
      </tr>

      {isOpen && team.members.map((member) => <Member key={member.id} date={date} theme={mainTheme} {...member} />)}
    </>
  )
}

export default React.memo(TeamComponent)
