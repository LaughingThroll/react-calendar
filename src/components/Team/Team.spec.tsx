import { shallow } from 'enzyme'
import Team, { TeamProps } from './'
import { normalTeams } from './../../__STUBS__/normalTeams'

const setUp = (props: TeamProps) => shallow(<Team {...props} />)

describe('Team render', () => {
  it('default', () => {
    const props = { date: new Date('2021-08-05'), team: normalTeams[0] }
    const tree = setUp(props)

    expect(tree).toMatchSnapshot()
  })
})
