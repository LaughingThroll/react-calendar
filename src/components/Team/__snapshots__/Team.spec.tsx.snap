// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team render default 1`] = `
<Fragment>
  <tr
    className="
        calendar-body__row  
        melrose-theme 
        melrose-theme--background 
         "
  >
    <td
      className="team"
    >
      <Memo(TeamInfo)
        date={2021-08-05T00:00:00.000Z}
        id="asdksadowkd"
        isOpen={true}
        members={
          Array [
            Object {
              "id": "asdffegegege",
              "name": "Eric Meyer",
              "vacations": Array [
                Object {
                  "endDate": "2021-02-15",
                  "startDate": "2021-02-12",
                  "type": "Paid",
                },
                Object {
                  "endDate": "2021-04-29",
                  "startDate": "2021-04-23",
                  "type": "UnPaid",
                },
                Object {
                  "endDate": "2021-04-21",
                  "startDate": "2021-04-14",
                  "type": "UnPaid",
                },
                Object {
                  "endDate": "2021-04-08",
                  "startDate": "2021-04-01",
                  "type": "Paid",
                },
              ],
            },
            Object {
              "id": "fllfhllkjj",
              "name": "Chris Coyier",
              "vacations": Array [
                Object {
                  "endDate": "2021-02-19",
                  "startDate": "2021-01-19",
                  "type": "Paid",
                },
                Object {
                  "endDate": "2021-02-22",
                  "startDate": "2021-02-20",
                  "type": "UnPaid",
                },
                Object {
                  "endDate": "2021-04-28",
                  "startDate": "2021-04-18",
                  "type": "UnPaid",
                },
                Object {
                  "endDate": "2021-04-01",
                  "startDate": "2021-03-29",
                  "type": "Paid",
                },
              ],
            },
            Object {
              "id": "sdaasdwdd",
              "name": "Jack London",
              "vacations": Array [
                Object {
                  "endDate": "2021-02-01",
                  "startDate": "2021-01-28",
                  "type": "Paid",
                },
                Object {
                  "endDate": "2021-01-18",
                  "startDate": "2021-01-15",
                  "type": "Paid",
                },
              ],
            },
            Object {
              "id": "wdwrtettete",
              "name": "Paul Irish",
              "vacations": Array [
                Object {
                  "endDate": "2021-02-22",
                  "startDate": "2021-02-20",
                  "type": "Paid",
                },
                Object {
                  "endDate": "2021-03-25",
                  "startDate": "2021-02-25",
                  "type": "Paid",
                },
                Object {
                  "endDate": "2021-04-28",
                  "startDate": "2021-04-18",
                  "type": "Paid",
                },
              ],
            },
          ]
        }
        name="Frontend Team"
        percentageOfAbsent={
          Array [
            0,
            2,
            0,
            0,
            1,
            22,
            2,
            2,
            2,
            2,
            11,
            1,
          ]
        }
        toggleOpen={[Function]}
      />
    </td>
    <Memo(Cell)
      cellDate={2021-07-31T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="0"
    />
    <Memo(Cell)
      cellDate={2021-08-01T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="1"
    />
    <Memo(Cell)
      cellDate={2021-08-02T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="2"
    />
    <Memo(Cell)
      cellDate={2021-08-03T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="3"
    />
    <Memo(Cell)
      cellDate={2021-08-04T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="4"
    />
    <Memo(Cell)
      cellDate={2021-08-05T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="5"
    />
    <Memo(Cell)
      cellDate={2021-08-06T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="6"
    />
    <Memo(Cell)
      cellDate={2021-08-07T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="7"
    />
    <Memo(Cell)
      cellDate={2021-08-08T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="8"
    />
    <Memo(Cell)
      cellDate={2021-08-09T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="9"
    />
    <Memo(Cell)
      cellDate={2021-08-10T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="10"
    />
    <Memo(Cell)
      cellDate={2021-08-11T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="11"
    />
    <Memo(Cell)
      cellDate={2021-08-12T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="12"
    />
    <Memo(Cell)
      cellDate={2021-08-13T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="13"
    />
    <Memo(Cell)
      cellDate={2021-08-14T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="14"
    />
    <Memo(Cell)
      cellDate={2021-08-15T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="15"
    />
    <Memo(Cell)
      cellDate={2021-08-16T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="16"
    />
    <Memo(Cell)
      cellDate={2021-08-17T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="17"
    />
    <Memo(Cell)
      cellDate={2021-08-18T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="18"
    />
    <Memo(Cell)
      cellDate={2021-08-19T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="19"
    />
    <Memo(Cell)
      cellDate={2021-08-20T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="20"
    />
    <Memo(Cell)
      cellDate={2021-08-21T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="21"
    />
    <Memo(Cell)
      cellDate={2021-08-22T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="22"
    />
    <Memo(Cell)
      cellDate={2021-08-23T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="23"
    />
    <Memo(Cell)
      cellDate={2021-08-24T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="24"
    />
    <Memo(Cell)
      cellDate={2021-08-25T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="25"
    />
    <Memo(Cell)
      cellDate={2021-08-26T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="26"
    />
    <Memo(Cell)
      cellDate={2021-08-27T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="27"
    />
    <Memo(Cell)
      cellDate={2021-08-28T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="28"
    />
    <Memo(Cell)
      cellDate={2021-08-29T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="29"
    />
    <Memo(Cell)
      cellDate={2021-08-30T21:00:00.000Z}
      classNames={
        Array [
          "calendar-body__cell",
        ]
      }
      key="30"
    />
    <Memo(Cell)
      classNames={
        Array [
          "calendar-body__cell",
          "cell-gray",
        ]
      }
    />
  </tr>
  <Memo(MemberComponent)
    date={2021-08-05T00:00:00.000Z}
    id="asdffegegege"
    key="asdffegegege"
    name="Eric Meyer"
    theme="melrose-theme"
    vacations={
      Array [
        Object {
          "endDate": "2021-02-15",
          "startDate": "2021-02-12",
          "type": "Paid",
        },
        Object {
          "endDate": "2021-04-29",
          "startDate": "2021-04-23",
          "type": "UnPaid",
        },
        Object {
          "endDate": "2021-04-21",
          "startDate": "2021-04-14",
          "type": "UnPaid",
        },
        Object {
          "endDate": "2021-04-08",
          "startDate": "2021-04-01",
          "type": "Paid",
        },
      ]
    }
  />
  <Memo(MemberComponent)
    date={2021-08-05T00:00:00.000Z}
    id="fllfhllkjj"
    key="fllfhllkjj"
    name="Chris Coyier"
    theme="melrose-theme"
    vacations={
      Array [
        Object {
          "endDate": "2021-02-19",
          "startDate": "2021-01-19",
          "type": "Paid",
        },
        Object {
          "endDate": "2021-02-22",
          "startDate": "2021-02-20",
          "type": "UnPaid",
        },
        Object {
          "endDate": "2021-04-28",
          "startDate": "2021-04-18",
          "type": "UnPaid",
        },
        Object {
          "endDate": "2021-04-01",
          "startDate": "2021-03-29",
          "type": "Paid",
        },
      ]
    }
  />
  <Memo(MemberComponent)
    date={2021-08-05T00:00:00.000Z}
    id="sdaasdwdd"
    key="sdaasdwdd"
    name="Jack London"
    theme="melrose-theme"
    vacations={
      Array [
        Object {
          "endDate": "2021-02-01",
          "startDate": "2021-01-28",
          "type": "Paid",
        },
        Object {
          "endDate": "2021-01-18",
          "startDate": "2021-01-15",
          "type": "Paid",
        },
      ]
    }
  />
  <Memo(MemberComponent)
    date={2021-08-05T00:00:00.000Z}
    id="wdwrtettete"
    key="wdwrtettete"
    name="Paul Irish"
    theme="melrose-theme"
    vacations={
      Array [
        Object {
          "endDate": "2021-02-22",
          "startDate": "2021-02-20",
          "type": "Paid",
        },
        Object {
          "endDate": "2021-03-25",
          "startDate": "2021-02-25",
          "type": "Paid",
        },
        Object {
          "endDate": "2021-04-28",
          "startDate": "2021-04-18",
          "type": "Paid",
        },
      ]
    }
  />
</Fragment>
`;
